### PR TITLE
Update manure excretion attributes in animal module

### PR DESCRIPTION
**Pull request by ltp23**
Hi reviewers, 

This PR mainly focuses on reorganizing the code for the manure calculations in the animal module and reviewing the formulas for those calculations. There is more work to be done to resolve the problem with negative values associated with nitrogen calculations. As Haowen suggested, the values for methane emission seem lower than expected. So, upcoming PRs should address these two main concerns.

Another refactoring PR would be helpful too since the manure calculations have similar structures across the four animal groups. 

Here is the link to the pseudocode [document](https://docs.google.com/document/d/1TNpOHkD4j_StRPHmMcaWoQ0ZFfigdM4L/edit).

### CODE REVIEW

**Comment by KFosterReed** [comment]
I just pushed a commit to fix the units... I am 99% sure they are correct now!

**Review by tomhuhh — APPROVED** [review]
LGTM!

File: RUFAS/routines/animal/manure/dry_cow_manure_excretion.py [hunk 1]
```diff
@@ -6,96 +6,111 @@
...
-    manure = 0.022 * bw + 21.844
+    total_manure_excreted = 0.022 * body_weight + 21.844
 
     # Total solids, kg/d [A.3D.A.2]
-    total_solids = 0.178 * dm_intake + 2.733
+    total_solids = 0.178 * dry_matter_intake + 2.733
 
     # Nitrogen in liquid and solid manure , g [A.3D.B.1]
-    N_manure = 12.747 * dm_intake + 1606.290 * CP_conc/100 - 117.5
+    nitrogen = 12.747 * dry_matter_intake + 1606.290 * CP_concentration / 100 - 117.5  # TODO: Divide by 1000?
```
Comment (tomhuhh) [inline RUFAS/routines/animal/manure/dry_cow_manure_excretion.py]: If you are looking for an answer with a unit of kg, yes you will want to divide by 1000. Thank you!

File: RUFAS/routines/animal/manure/calf_manure_excretion.py [hunk 2]
```diff
@@ -6,64 +6,98 @@
...
+    # Non-degradable volatile solids, kg/day
+    non_degradable_volatile_solids = total_volatile_solids - degradable_volatile_solids
+
+    # Nitrogen excretion
+    # Amount of nitrogen excreted by the calf, kg [A.3A.B.1]
+    # TODO: Review this part. It was added to get the dry matter intake and CP concentration.
+    nutrient_amounts, nutrient_concentrations = ration_report(ration_formulation, feed.available_feeds)
+    dry_matter_intake = nutrient_amounts['dm']
+    CP_concentration = nutrient_concentrations['CP']
+    manure_nitrogen = 112.55 * dry_matter_intake * (CP_concentration / 100)
```
Comment (tomhuhh) [inline RUFAS/routines/animal/manure/calf_manure_excretion.py:69]: You want to divide it by 1000 to convert the output to kg. Thanks!

File: RUFAS/routines/animal/manure/lactating_cow_manure_excretion.py [hunk 3]
```diff
@@ -6,135 +6,221 @@
...
 import math
 
+from RUFAS.general_constants import GeneralConstants
+from RUFAS.routines.animal.manure.general_manure import AnimalManureExcretions
+from RUFAS.routines.animal.manure.general_manure import calculate_phosphorus_excretion_values
+from RUFAS.routines.animal.ration.ration_driver import ration_report
+
+
+def manure_calculations(ration_formulation,
+                        feed,
```
Comment (tomhuhh) [inline RUFAS/routines/animal/manure/lactating_cow_manure_excretion.py:19]: As Joe mentioned while I was talking to him, a suggestion here is to label "ration_formulation" and "feed" as dictionary, for example, "feed: dict". As this is not my area of expertise, I will leave it to your decision. Thank you!

File: RUFAS/routines/animal/manure/growing_heifer_manure_excretion.py [hunk 4]
```diff
@@ -6,80 +6,126 @@
...
+    fecal_nitrogen = manure_nitrogen - urine_nitrogen  # TODO: Unused
+
+    # Nitrogen concentration in urinary urea, g urea-N/L [A.3G.B.1]
+    urinary_nitrogen_concentration = (urine_nitrogen *GeneralConstants.KG_TO_GRAMS) / urine
+    urine_urea_nitrogen_concentration = -1.16 + 0.86 * urinary_nitrogen_concentration
+
+    # Total ammoniacal nitrogen concentration in the manure slurry,
+    # g ammoniacal nitrogen/L manure slurry [A.3G.B.3]
+    tan_percent_of_urea = 48.2 - 2.9 * urine_urea_nitrogen_concentration
+    total_ammoniacal_nitrogen_concentration = (tan_percent_of_urea / 100) * urine_urea_nitrogen_concentration
```
Comment (tomhuhh) [inline RUFAS/routines/animal/manure/growing_heifer_manure_excretion.py:93]: Just wanted to make a note here. The ammoniacal N conc is negative. This should have already been flagged by others. 
Screenshot 2023-03-08 at 3 57 33 PM

File: RUFAS/routines/animal/manure/growing_heifer_manure_excretion.py [hunk 5]
```diff
@@ -6,80 +6,126 @@
...
+
+    # Amount of potassium excreted, g [A.3D.B.3]
+    potassium = dry_matter_intake * (potassium_concentration / 100) * GeneralConstants.KG_TO_GRAMS
+
+    # Methane emissions, g/day [A.3B.C.1]
+    # Methane model = Boadi
+    methane_emission = (38.62 + 26.44 * dry_matter_intake) * 0.554
```
Comment (tomhuhh) [inline RUFAS/routines/animal/manure/growing_heifer_manure_excretion.py:100]: Methane emissions are a little low (~250 g/d) for HeiferIII (pen 2). Might need some investigation. Thanks! 
Screenshot 2023-03-08 at 4 03 11 PM

File: RUFAS/routines/animal/manure/lactating_cow_manure_excretion.py [hunk 6]
```diff
@@ -6,135 +6,221 @@
...
+    # Nitrogen concentration in urinary urea, g urea-N/L [A.3G.B.1]
+    urinary_nitrogen_concentration = (urine_nitrogen * GeneralConstants.KG_TO_GRAMS) / urine
+    urine_urea_nitrogen_concentration = -1.16 + 0.86 * urinary_nitrogen_concentration
 
     # Total ammoniacal nitrogen concentration in the manure slurry,
-    # mol/L (Eq 6.1)
-    TAN_s = (-162.4 * U * U + 96.4 * U) / 100
+    # g ammoniacal nitrogen/L manure slurry [A.3G.B.3]
+    tan_percent_of_urea = 48.2 - 2.9 * urine_urea_nitrogen_concentration
+    total_ammoniacal_nitrogen_concentration = (tan_percent_of_urea / 100) * urine_urea_nitrogen_concentration
```
Comment (tomhuhh) [inline RUFAS/routines/animal/manure/lactating_cow_manure_excretion.py:147]: Wanted to make a note here that ammoniacal N conc are negatives. This should probably have been flagged. 
Screenshot 2023-03-08 at 4 16 31 PM

File: RUFAS/routines/animal/manure/general_manure.py [hunk 7]
```diff
@@ -1,37 +1,113 @@
...
-
-    # manure P excretion for manure module input (g) (A.3.B.2)
-    p_excrt_manure = p_feces_excrt + p_urine
-    
-    # amount of P excreted by an animal (g) (A.3.B.3)
-    p_excrt = p_milk + p_feces_excrt + p_urine
-    
-    return p_excrt, WIP_frac, WOP_frac, p_excrt_manure, p_frac
+    # compromised of organic water extractable P [A.3.A.3]
+    # TODO: Is it 0.05 or 0.50?
```
Comment (tomhuhh) [inline RUFAS/routines/animal/manure/general_manure.py:100]: I checked the original paper that was cited in the pseudocode but could not find a good explanation for it. Keep in mind this is only the water extractable section of the total manure phosphorus. The coefficients do not need to add up to 1. Thanks!

**Review by varrma — APPROVED** [review]

**Review by tomhuhh — APPROVED** [review]

**Review by tomhuhh — APPROVED** [review]

File: RUFAS/routines/animal/manure/calf_manure_excretion.py [hunk 8]
```diff
@@ -6,64 +6,98 @@
...
+    body_weight : float
+        Body weight of the current animal, kg.
+    fecal_phosphorus : float
+        Amount of fecal phosphorus excreted by the current animal, g.
+    urine_phosphorus_required : float
+        Amount of phosphorus required for urine production, g.
+
+    Returns
+    -------
+    float
```
Comment (ew3361zh) [inline RUFAS/routines/animal/manure/calf_manure_excretion.py:39]: Do we need to include the variable names on the returns as well?

```suggestion
    total_phosphorus_excreted : float
```

File: RUFAS/routines/animal/manure/dry_cow_manure_excretion.py [hunk 9]
```diff
@@ -6,96 +6,154 @@
...
+    fecal_phosphorus : float
+        Amount of fecal phosphorus excreted by the current animal, g.
+    urine_phosphorus_required : float
+        Amount of phosphorus required for urine production, g.
+    metabolizable_energy_intake : float
+        Metabolizable energy intake, Mcal/kg dry matter.
+
+    Returns
+    -------
+    float
```
Comment (ew3361zh) [inline RUFAS/routines/animal/manure/dry_cow_manure_excretion.py:47]: does this need the variable name it's returning before the type?
```suggestion
    total_phosphorus_excreted : float
```

File: RUFAS/routines/animal/manure/lactating_cow_manure_excretion.py [hunk 10]
```diff
@@ -6,135 +6,201 @@
...
+    methane_model : str
+        Methane model used for methane emission calculations, including Mutian, Mills, IPCC.
+    milk_fat : float
+        Milk fat (from animal input), % of milk.
+    metabolizable_energy_intake : float
+        Metabolizable energy intake, Mcal/kg dry matter.
+
+    Returns
+    -------
+    float
```
Comment (ew3361zh) [inline RUFAS/routines/animal/manure/lactating_cow_manure_excretion.py:59]: Similar to other comments, do the returns need to have their variable names included?
```suggestion
    total_phosphorus_excreted : float
```

**Review by ew3361zh — APPROVED** [review]
Very nice work, Loi. This PR is heavy on scientific terms and equations and I will assume from the multiple SME reviews that the formulas, descriptions, variable names, output values, and specific hard-coded values are correct.

Code wise, I had one comment about return variable doc strings that came up in a few places but I ran this with all the different input files with no issue and the pytests all passed.

**Pull request merged**